libcontainer/container_linux.go

```diff
@@ -431,6 +435,81 @@ func (c *Container) signal(s os.Signal) error {
 	return nil
 }
 
+func (c *Container) killViaPidfd() error {
+	pidfd, err := unix.PidfdOpen(c.initProcess.pid(), 0)
+	if err != nil {
+		return err
+	}
+	defer unix.Close(pidfd)
+
+	epollfd, err := unix.EpollCreate1(unix.EPOLL_CLOEXEC)
+	if err != nil {
+		return err
+	}
+	defer unix.Close(epollfd)
+
+	event := unix.EpollEvent{
+		Events: unix.EPOLLIN,
+		Fd:     int32(pidfd),
+	}
+	if err := unix.EpollCtl(epollfd, unix.EPOLL_CTL_ADD, pidfd, &event); err != nil {
+		return err
+	}
+
+	if err := unix.PidfdSendSignal(pidfd, unix.SIGKILL, nil, 0); err != nil {
+		return err
+	}
+
+	events := make([]unix.EpollEvent, 1)
+	for {
+		// Set the timeout to 10s, the same as in kill below.
+		n, err := unix.EpollWait(epollfd, events, 10000)
+		if err != nil {
+			if err == unix.EINTR {
+				continue
+			}
+			return err
+		}
+
+		if n == 0 {
+			return errors.New("container init still running")
+		}
+
+		if n > 0 {
+			event := events[0]
+			if event.Fd == int32(pidfd) {
+				return nil
+			}
+		}
+	}
+}
+
+func (c *Container) kill() error {
+	_ = c.Signal(unix.SIGKILL)
+	for i := 0; i < 100; i++ {
+		time.Sleep(100 * time.Millisecond)
+		if err := c.Signal(unix.Signal(0)); err != nil {
+			return nil
+		}
+	}
+	return errors.New("container init still running")
+}
+
+// EnsureKilled kills the container and waits for the kernel to finish killing it.
+func (c *Container) EnsureKilled() error {
+	// When a container doesn't have a private pidns, we have to kill all processes
+	// in the cgroup, it's more simpler to use `cgroup.kill` or `unix.Kill`.
+	if c.config.Namespaces.IsPrivate(configs.NEWPID) {
```

Comment:
Sorry, I don't get this. I've seen the resolved comment. I'm sure it makes sense if it makes sense for 2 persons.

But we are not using cgroup.kill here, right? Why not use it?

Currently if the container uses pidns then we use pidfd to kill it, but those two things seem completely unrelated to me. My mental mode is:
1. Use cgroup.kill if the kernel supports it
1. Use pidfd if the kernel supports it
1. Just send a signal otherwise

But this is mixing if the container is using pidns with using pidfd. Why wouldn't we use pidfd to kill the process if cgroup.kill is not supported and the kernel supports pidfd?

Reply:
> But we are not using cgroup.kill here, right? Why not use it?

If the container doesn't have a private pid ns, it will use `signalAllProcesses` to kill all the processes in the cgroup.
https://github.com/opencontainers/runc/blob/df920a9c24a1916146335a3528e48a915dc1a395/libcontainer/init_linux.go#L663-L678

Reply:
Ohhh, I see this comment: https://github.com/opencontainers/runc/blob/c3a41d77db71d9ba8bd76cdcd74d8aa71d16d2a8/libcontainer/container_linux.go#L387-L393

Okay, now I get it!